src/inc/stackstring.h

```diff
@@ -0,0 +1,196 @@
+        {
+            SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+
+            DeleteBuffer();
+            m_count = 0;
+
+            return;
+        }
+
+        DeleteBuffer();
+        m_buffer = newBuffer;
+        m_count = count;
+        m_size = count+1;
+
+        return;
+    }
+
+    void Resize(SIZE_T count)
+    {
+        if (NULL == m_buffer)
+        {
+            if (count > STACKCOUNT)
+            {
+                ReallocateBuffer(count);
+            }
+            else
+            {
+                m_size = STACKCOUNT+1;
+                m_buffer = m_innerBuffer;
+                m_count = count;
+            }
+        }
+        else if (m_innerBuffer == m_buffer)
+        {
+            if (count > STACKCOUNT)
+            {
+                ReallocateBuffer(count);
+            }
+            else
+            {
+                m_count = count;
+                m_size = STACKCOUNT+1;
+            }
+        }
+        else
+        {
+            ReallocateBuffer(count);
+        }
+
+        return;
+    }
+
+    StackString(const StackString &s)
+    {
+        Set(s);
+    }
+
+public:
+    StackString()
+        : m_buffer(m_innerBuffer), m_count(0)
+    {
+    }
+
+    BOOL Set(const T * buffer, SIZE_T count)
+    {
+        Resize(count);
+        if (NULL == m_buffer)
+            return FALSE;
+
+        CopyMemory(m_buffer, buffer, (count + 1) * sizeof(T));
+        NullTerminate();
+        return TRUE;
+    }
+
+    BOOL Set(const StackString &s)
+    {
+        return Set(s.m_buffer, s.m_count);
+    }
+
+    SIZE_T GetCount() const
+    {
+        return m_count;
+    }
+
+    SIZE_T GetAllocationCount() const
+    {
+        return m_size;
+    }
+    
+    SIZE_T GetSizeOf() const
+    {
+        return m_size * sizeof(T);
+    }
+
+    CONST T * GetString() const
+    {
+        return (const T *)m_buffer;
+    }
+
+    T * ExtractString() 
+    {
+        T * ret_buffer;
+
+        if (m_innerBuffer != m_buffer)
+        {
+            //it is heap allocated just return the memory
+            ret_buffer = m_buffer;
+            m_buffer = NULL;
+        }
+        else
+        {
+            ret_buffer = (T *)malloc((m_count + 1) * sizeof(T));
+            if (NULL == ret_buffer)
+            {
+                SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+                return NULL;
+            }
+            CopyMemory(ret_buffer, m_buffer, (m_count + 1) * sizeof(T));
+            ret_buffer[m_count] = 0;
+        }
+        return ret_buffer;
+    }
+
+    T * OpenStringBuffer(SIZE_T count)
+    {
+        Resize(count);
+        return (T *)m_buffer;
+    }
+
+    void Clear()
+    {
+        m_count = 1;
+        NullTerminate();
+        return;
+    }
+    void CloseBuffer(SIZE_T count)
+    {
+        if (m_count > count)
+            m_count = count;
+
+        NullTerminate();
+        return;
+    }
+    
+    ~StackString()
+    {
+        DeleteBuffer();
+    }
+};
+
+#if _DEBUG
```

Comment:
Are we using the following typedefs in the VM or is it just a copy-paste from PAL?